core/thread/inc/TPool.h

```diff
@@ -0,0 +1,204 @@
+   template<class F, class INTEGER, class R> auto MapReduce(F func, ROOT::TSeq<INTEGER> args, R redfunc) -> decltype(func(args.front()));
+   template<class F, class T, class R> auto MapReduce(F func, T &args, R redfunc) -> decltype(++(args.begin()), args.end(), func(args.front()));
+   // /// \cond doxygen should ignore these methods
+   template<class F, class R> auto MapReduce(F func, TCollection &args, R redfunc) -> decltype(func(nullptr));
+   template<class F, class T, class R> auto MapReduce(F func, std::initializer_list<T> args, R redfunc) -> decltype(func(*args.begin()));
+   template<class F, class T, class R> auto MapReduce(F func, std::vector<T> &args, R redfunc) -> decltype(func(args.front()));
+   // /// \endcond
+
+protected:
+   template<class T, class R> auto Reduce(const std::vector<T> &objs, R redfunc) -> decltype(redfunc(objs));
+   template<class T, class BINARYOP> auto Reduce(const std::vector<T> &objs, BINARYOP redfunc) -> decltype(redfunc(objs.front(), objs.front()));
+};
+
+//////////////////////////////////////////////////////////////////////////
+/// Execute func (with no arguments) nTimes in parallel.
+/// A vector containg executions' results is returned.
+/// Functions that take more than zero arguments can be executed (with
+/// fixed arguments) by wrapping them in a lambda or with std::bind.
+template<class subc> template<class F>
+auto TPool<subc>::Map(F func, unsigned nTimes) -> std::vector<decltype(func())>
+{
+   return static_cast<subc*>(this)->Map(func, nTimes);
+}
+
+// //////////////////////////////////////////////////////////////////////////
+// /// Execute func in parallel distributing the elements of the args collection between the workers.
+// /// See class description for the valid types of collections and containers that can be used.
+// /// A vector containing each execution's result is returned. The user is responsible of deleting
+// /// objects that might be created upon the execution of func, returned objects included.
+// /// **Note:** the collection of arguments is modified by Map and should be considered empty or otherwise
+// /// invalidated after Map's execution (std::move might be applied to it).
+template<class subc> template<class F, class T>
+auto TPool<subc>::Map(F func, T &args) -> std::vector < decltype(++(args.begin()), args.end(), func(args.front())) >
+{
+   std::vector<typename T::value_type> vargs(
+      std::make_move_iterator(std::begin(args)),
+      std::make_move_iterator(std::end(args))
+   );
+   const auto &reslist = static_cast<subc*>(this)->Map(func, vargs);
+   return reslist;
+}
+
+
+// tell doxygen to ignore this (\endcond closes the statement)
+/// \cond
+template<class subc> template<class F, class INTEGER>
+auto TPool<subc>::Map(F func, ROOT::TSeq<INTEGER> args) -> std::vector<decltype(func(args.front()))>
+{
+  return static_cast<subc*>(this)->Map(func, args);
+}
+
+template<class subc> template<class F>
+TObjArray TPool<subc>::Map(F func, TCollection &args)
```

Comment:
I think the right solution is to use a free function version of size (this is where the C++ standard is going: http://en.cppreference.com/w/cpp/iterator/size) where you can use size or Size or even sizeof for arrays.

For value_type ... either also a free const_expr functions/template and/or using decl_type of *begin